app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt

```diff
@@ -111,9 +112,27 @@ class BrowserWebViewClient(
     @WorkerThread
     override fun shouldInterceptRequest(webView: WebView, request: WebResourceRequest): WebResourceResponse? {
         return runBlocking {
+
             val documentUrl = withContext(Dispatchers.Main) { webView.url }
             Timber.v("Intercepting resource ${request.url} on page $documentUrl")
-            requestInterceptor.shouldIntercept(request, webView, documentUrl, webViewClientListener)
+            val interceptionAction = requestInterceptor.shouldIntercept(request, webView, documentUrl, webViewClientListener)
+
+            if (interceptionAction is InterceptAction.BlockAndLoadAlternativeUrl) {
+                withContext(Dispatchers.Main) {
+                    webView.loadUrl(interceptionAction.url.toString())
+                }
+            }
+
+            return@runBlocking convertActionToWebResourceResponse(interceptionAction)
+        }
+    }
+
+    private fun convertActionToWebResourceResponse(action: InterceptAction): WebResourceResponse? {
```

Comment:
This api doesn't read well for me - why does the InterceptAction sometime return the WebResourceResponse and not others? I feel like it should always provide it or never,